Introduce `apm-server.auth.*` config

axw opened the pull request:
## Motivation/summary

Introduce the new AgentAuth config structure, which holds API Key and secret token auth. Later we will add "anonymous" auth here too, and deprecate/replace some RUM config (rate limiting and allowed service names).

We also introduce a new YAML naming scheme for the config, `apm-server.auth.*`. The old config is deprecated and copied across to the new config fields.

## Checklist

- [x] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)
- [x] Documentation has been updated

## How to test these changes

1. Set `apm-server.secret_token` and `apm-server.api_key.*`, make sure they are honoured (e.g. query "GET /" with/out auth)
2. Set `apm-server.auth.secret_token` and `apm-server.auth.api_key.*`, same again.
3. Set both new and old with different values: check that a warning is logged that the old config is ignored, and check that it is possible to auth using the new but not old

## Related issues

https://github.com/elastic/apm-server/issues/5347

apmmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-5457/detail/PR-5457/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-5457/detail/PR-5457/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-5457/detail/PR-5457/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-5457/detail/PR-5457/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://null_5457.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats


* Build Cause: `Pull request #5457 opened`



* Start Time: 2021-06-15T10:04:40.220+0000


* Duration: 4 min 43 sec


* Commit: 8647dda1f75da317beba525ac06f3017aadbdf43




#### Trends :test_tube:


![Image of Build Times](https://apm-ci.elastic.co/job/apm-server/job/apm-server-mbp/job/PR-5457/buildTimeGraph/png)



</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->









### Steps errors [![3](https://img.shields.io/badge/3%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-5457/detail/PR-5457/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Check out from version control`
<ul>
<li>Took 0 min 8 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-server/pipelines/apm-server-mbp/pipelines/PR-5457/runs/1/steps/11/log/?start=0">here</a></li>
<li>Description: <code>[2021-06-15T10:07:10.757Z] The recommended git tool is: git
[2021-06-15T10:07:12.958Z] using creden</code></l1>
</ul>
  
  
  
  
##### `Check out from version control`
<ul>
<li>Took 0 min 5 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-server/pipelines/apm-server-mbp/pipelines/PR-5457/runs/1/steps/15/log/?start=0">here</a></li>
<li>Description: <code>[2021-06-15T10:07:39.012Z] The recommended git tool is: git
[2021-06-15T10:07:39.021Z] using creden</code></l1>
</ul>
  
  
  
  
##### `Check out from version control`
<ul>
<li>Took 0 min 7 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-server/pipelines/apm-server-mbp/pipelines/PR-5457/runs/1/steps/19/log/?start=0">here</a></li>
<li>Description: <code>[2021-06-15T10:08:14.612Z] The recommended git tool is: git
[2021-06-15T10:08:14.634Z] using creden</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-06-15T10:07:39.083Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-06-15T10:07:39.088Z]  > git fetch --no-tags --progress -- git@github.com:elastic/apm-server.git +refs/heads/*:refs/remotes/origin/* # timeout=15
[2021-06-15T10:07:43.720Z] Cleaning workspace
[2021-06-15T10:07:43.733Z] Using shallow fetch with depth 3
[2021-06-15T10:07:43.733Z] Pruning obsolete local branches
[2021-06-15T10:07:44.438Z] Merging remotes/origin/master commit f7c7b0c6873d40f533c072b3d9751c8fe285e63e into PR head commit 8647dda1f75da317beba525ac06f3017aadbdf43
[2021-06-15T10:07:44.468Z] ERROR: Execution failed
[2021-06-15T10:07:44.468Z] hudson.plugins.git.GitException: Command "git checkout -f 8647dda1f75da317beba525ac06f3017aadbdf43" returned status code 128:
[2021-06-15T10:07:44.468Z] stdout: 
[2021-06-15T10:07:44.468Z] stderr: fatal: reference is not a tree: 8647dda1f75da317beba525ac06f3017aadbdf43
[2021-06-15T10:07:44.468Z] 
[2021-06-15T10:07:44.468Z] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2639)
[2021-06-15T10:07:44.468Z] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$1100(CliGitAPIImpl.java:84)
[2021-06-15T10:07:44.468Z] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$9.execute(CliGitAPIImpl.java:2956)
[2021-06-15T10:07:44.468Z] Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to JNLP4-connect connection from apm-ci-immutable-ubuntu-1804-1623751497245822780.c.elastic-ci-prod.internal/10.224.1.15:55916
[2021-06-15T10:07:44.468Z] 		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1800)
[2021-06-15T10:07:44.468Z] 		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
[2021-06-15T10:07:44.468Z] 		at hudson.remoting.Channel.call(Channel.java:1001)
[2021-06-15T10:07:44.468Z] 		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:146)
[2021-06-15T10:07:44.468Z] 		at sun.reflect.GeneratedMethodAccessor799.invoke(Unknown Source)
[2021-06-15T10:07:44.468Z] 		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[2021-06-15T10:07:44.468Z] 		at java.lang.reflect.Method.invoke(Method.java:498)
[2021-06-15T10:07:44.468Z] 		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:132)
[2021-06-15T10:07:44.468Z] 		at com.sun.proxy.$Proxy136.execute(Unknown Source)
[2021-06-15T10:07:44.468Z] 		at jenkins.plugins.git.MergeWithGitSCMExtension.checkout(MergeWithGitSCMExtension.java:145)
[2021-06-15T10:07:44.468Z] 		at jenkins.plugins.git.MergeWithGitSCMExtension.decorateRevisionToBuild(MergeWithGitSCMExtension.java:111)
[2021-06-15T10:07:44.468Z] 		at hudson.plugins.git.GitSCM.determineRevisionToBuild(GitSCM.java:1163)
[2021-06-15T10:07:44.468Z] 		at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1302)
[2021-06-15T10:07:44.468Z] 		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:125)
[2021-06-15T10:07:44.468Z] 		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:93)
[2021-06-15T10:07:44.468Z] 		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:80)
[2021-06-15T10:07:44.468Z] 		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
[2021-06-15T10:07:44.468Z] 		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[2021-06-15T10:07:44.468Z] 		at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[2021-06-15T10:07:44.468Z] 		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2021-06-15T10:07:44.468Z] 		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2021-06-15T10:07:44.468Z] Caused: hudson.plugins.git.GitException: Could not checkout 8647dda1f75da317beba525ac06f3017aadbdf43
[2021-06-15T10:07:44.468Z] 	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$9.execute(CliGitAPIImpl.java:2980)
[2021-06-15T10:07:44.468Z] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:161)
[2021-06-15T10:07:44.468Z] 	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:154)
[2021-06-15T10:07:44.468Z] 	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
[2021-06-15T10:07:44.468Z] 	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
[2021-06-15T10:07:44.468Z] 	at hudson.remoting.Request$2.run(Request.java:375)
[2021-06-15T10:07:44.468Z] 	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:73)
[2021-06-15T10:07:44.468Z] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[2021-06-15T10:07:44.468Z] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2021-06-15T10:07:44.468Z] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2021-06-15T10:07:44.468Z] 	at hudson.remoting.Engine$1.lambda$newThread$0(Engine.java:118)
[2021-06-15T10:07:44.468Z] 	at java.lang.Thread.run(Thread.java:748)
[2021-06-15T10:07:44.468Z] Retrying
[2021-06-15T10:07:43.694Z]  > git config remote.origin.url git@github.com:elastic/apm-server.git # timeout=10
[2021-06-15T10:07:43.703Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2021-06-15T10:07:43.711Z]  > git config remote.origin.url git@github.com:elastic/apm-server.git # timeout=10
[2021-06-15T10:07:43.722Z]  > git rev-parse --verify HEAD # timeout=10
[2021-06-15T10:07:43.726Z] No valid HEAD. Skipping the resetting
[2021-06-15T10:07:43.726Z]  > git clean -fdx # timeout=10
[2021-06-15T10:07:43.735Z] Fetching upstream changes from git@github.com:elastic/apm-server.git
[2021-06-15T10:07:43.736Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-06-15T10:07:43.739Z]  > git fetch --no-tags --progress --prune -- git@github.com:elastic/apm-server.git +refs/pull/5457/head:refs/remotes/origin/PR-5457 +refs/heads/master:refs/remotes/origin/master # timeout=15
[2021-06-15T10:07:44.440Z]  > git config core.sparsecheckout # timeout=10
[2021-06-15T10:07:44.446Z]  > git checkout -f 8647dda1f75da317beba525ac06f3017aadbdf43 # timeout=15
[2021-06-15T10:07:44.549Z] Sleeping for 30 sec
[2021-06-15T10:08:14.612Z] The recommended git tool is: git
[2021-06-15T10:08:14.634Z] using credential f6c7695a-671e-4f4f-a331-acdce44ff9ba
[2021-06-15T10:08:14.652Z] Wiping out workspace first.
[2021-06-15T10:08:14.688Z] Cloning the remote Git repository
[2021-06-15T10:08:14.688Z] Using shallow clone with depth 3
[2021-06-15T10:08:14.688Z] Avoid fetching tags
[2021-06-15T10:08:14.697Z] Cloning repository git@github.com:elastic/apm-server.git
[2021-06-15T10:08:14.697Z]  > git init /var/lib/jenkins/workspace/pm-server_apm-server-mbp_PR-5457 # timeout=10
[2021-06-15T10:08:14.702Z] Fetching upstream changes from git@github.com:elastic/apm-server.git
[2021-06-15T10:08:14.702Z]  > git --version # timeout=10
[2021-06-15T10:08:14.705Z]  > git --version # 'git version 2.17.1'
[2021-06-15T10:08:14.705Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-06-15T10:08:14.710Z]  > git fetch --no-tags --progress -- git@github.com:elastic/apm-server.git +refs/heads/*:refs/remotes/origin/* # timeout=15
[2021-06-15T10:08:19.794Z] Cleaning workspace
[2021-06-15T10:08:19.821Z] Using shallow fetch with depth 3
[2021-06-15T10:08:19.821Z] Pruning obsolete local branches
[2021-06-15T10:08:19.771Z]  > git config remote.origin.url git@github.com:elastic/apm-server.git # timeout=10
[2021-06-15T10:08:19.775Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2021-06-15T10:08:19.785Z]  > git config remote.origin.url git@github.com:elastic/apm-server.git # timeout=10
[2021-06-15T10:08:19.797Z]  > git rev-parse --verify HEAD # timeout=10
[2021-06-15T10:08:19.801Z] No valid HEAD. Skipping the resetting
[2021-06-15T10:08:19.801Z]  > git clean -fdx # timeout=10
[2021-06-15T10:08:19.823Z] Fetching upstream changes from git@github.com:elastic/apm-server.git
[2021-06-15T10:08:19.823Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-06-15T10:08:19.827Z]  > git fetch --no-tags --progress --prune -- git@github.com:elastic/apm-server.git +refs/pull/5457/head:refs/remotes/origin/PR-5457 +refs/heads/master:refs/remotes/origin/master # timeout=15
[2021-06-15T10:08:20.629Z] Merging remotes/origin/master commit f7c7b0c6873d40f533c072b3d9751c8fe285e63e into PR head commit 8647dda1f75da317beba525ac06f3017aadbdf43
[2021-06-15T10:08:20.631Z]  > git config core.sparsecheckout # timeout=10
[2021-06-15T10:08:20.634Z]  > git checkout -f 8647dda1f75da317beba525ac06f3017aadbdf43 # timeout=15
[2021-06-15T10:08:21.950Z] Running on Jenkins in /var/lib/jenkins/workspace/pm-server_apm-server-mbp_PR-5457
[2021-06-15T10:08:22.186Z] [39;49m[INFO] getVaultSecret: Getting secrets[0m
[2021-06-15T10:08:22.277Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-06-15T10:08:23.738Z] + chmod 755 generate-build-data.sh
[2021-06-15T10:08:23.738Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-server/apm-server-mbp/PR-5457/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-server/apm-server-mbp/PR-5457/runs/1 FAILURE 223236
[2021-06-15T10:08:23.738Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-server/apm-server-mbp/PR-5457/runs/1/steps/?limit=10000 -o steps-info.json
[2021-06-15T10:08:25.081Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-server/apm-server-mbp/PR-5457/runs/1/tests/?status=FAILED -o tests-errors.json
[2021-06-15T10:08:25.081Z] Retry 1/3 exited 22, retrying in 1 seconds...
[2021-06-15T10:08:25.993Z] Retry 2/3 exited 22, retrying in 2 seconds...
[2021-06-15T10:08:27.854Z] Retry 3/3 exited 22, no more retries left.

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

simitt commented on docs/configuration-process.asciidoc:
```diff
@@ -99,6 +103,8 @@ The token is not used for RUM endpoints. By default, no authorization token is s
 It is recommended to use an authorization token in combination with SSL enabled.
 Read more about <<securing-apm-server, Securing APM Server>> and the <<secret-token, secret token>>.
 
+deprecated::[7.14.0, "In versions prior to 7.14.0, secret token authorization was known as `apm-server.secret_token`. In 7.14.0 this was renamed `apm-server.auth.secret_Token`. The old configuration will continue to work until 8.0.0, and the new configuration will take precedence."]
```
typo: `secret_Token`

axw replied:
thanks, fixed

simitt reviewed (commented):
Looks great.

axw commented:
@bmorelli25 would you please take a look at the docs changes? I've renamed some config, and added new deprecation sections for the deprecated config names.

I've also moved "api_key.* configuration options" into the "API keys" section, not sure if it was intentional that it was separate? I can move back if you prefer, but then I'm not sure where the deprecation section should go.

simitt reviewed (changes requested):
Can you also change the `apmpackage` please. I suggest we remove the deprecated settings and only support the new ones.

axw commented:
@simitt will do. I was going to wait for your changes (https://github.com/elastic/apm-server/pull/5444) to land, but I'll just update now for the purposes of review and update again when that lands.

bmorelli25 reviewed (approved):
Docs look great–thanks!

It looks like there's one [additional change](https://github.com/elastic/apm-server/blame/master/docs/copied-from-beats/docs/command-reference.asciidoc/#L142-L143) that needs to be made to the command reference file. I'll fix that in the Beats repo and copy it over to apm-server before 7.14.

axw commented:
Failure is related to the beats update, which will be resolved by https://github.com/elastic/apm-server/pull/5471

simitt commented:
Tested with BC2:
* old settings work as expected
* new settings work as expected
* API Key auth disabled when:
```
apm-server.api_key.enabled: true
apm-server.auth.api_key.enabled: false
```
  * Secret token 'abcd' expected
```
apm-server.secret_token: 'xxx'
apm-server.auth.secret_token: 'abcd'
```
